components/script/dom/globalscope.rs

```diff
@@ -1254,23 +1253,64 @@ impl GlobalScope {
         &self,
         id: Uuid,
         promise: Rc<Promise>,
-        f: Box<dyn Fn(Rc<Promise>, Result<Vec<u8>, ()>) + Send>,
+        f: Box<dyn Fn(Rc<Promise>, Result<Vec<u8>, Error>) + Send>,
     ) {
         let recv = self.send_msg(id);
         let trusted_promise = TrustedPromise::new(promise);
         let canceller = self.task_canceller(TaskSourceName::FileReading);
         let task_source = self.file_reading_task_source();
-        thread::spawn(move || {
-            let bytes = GlobalScope::read_msg(recv);
-            let _ = task_source.queue_with_canceller(
-                task!(resolve_promise: move || {
-                    let promise = trusted_promise.root();
-                    let _ac = enter_realm(&*promise.global());
-                    f(promise, bytes);
-                }),
-                &canceller,
-            );
-        });
+
+        ROUTER.add_route(
+            recv.to_opaque(),
+            Box::new(move |message| {
+                let message = message.to();
+                let mut bytes = vec![];
+                
+                match message {
+                    Ok(ReadFileProgress::Meta(mut blob_buf)) => {
+                        bytes.append(&mut blob_buf.bytes);
+                    },
+                    Ok(ReadFileProgress::Partial(mut bytes_in)) => {
+                        bytes.append(&mut bytes_in);
+                    },
+                    Ok(ReadFileProgress::EOF) => {
+                        let bytes = Ok(bytes);
+                        let _ = task_source.queue_with_canceller(
+                            task!(resolve_promise: move || {
+                                let promise = trusted_promise.root();
```

Comment:
I think you need to clone the promise(outside of the `task!`), and then move the clone into the closure of the `task!`. You might want to implement a `struct` to give the code some more structure, see the examples of [`TimerListener`](https://github.com/servo/servo/blob/04b26069a57de61593c54d2d689bd9be75b79da6/components/script/dom/globalscope.rs#L283) and [`MessageListener`](https://github.com/servo/servo/blob/04b26069a57de61593c54d2d689bd9be75b79da6/components/script/dom/globalscope.rs#L311).